apps/src/dance/LottieDancerRenderer.ts

```diff
@@ -0,0 +1,673 @@
+
+    const paintNode = (shapeNode: LottieShapeAny, rgba: RGBA | null) => {
+      if (!rgba) return;
+      const ty = shapeNode.ty;
+      if (
+        (ty === 'fl' || ty === 'st') &&
+        (shapeNode as LottieShapeFillOrStroke).c
+      ) {
+        setColorNode((shapeNode as LottieShapeFillOrStroke).c, rgba);
+      }
+      // Note: gradient fill/stroke (gf/gs) not handled here (documented elsewhere).
+    };
+
+    const walkShapes = (
+      items: Array<LottieShapeAny> | undefined,
+      layerName: string,
+      pathNames: string[] = []
+    ) => {
+      if (!Array.isArray(items)) return;
+
+      const layerSegs = splitSegments(layerName);
+
+      for (const shapeNode of items) {
+        if (!shapeNode) continue;
+
+        // Recurse into groups to build context (“pathNames”)
+        if (shapeNode.ty === 'gr') {
+          const nm = (shapeNode as LottieShapeGroup).nm || '';
+          const nextPath = nm ? pathNames.concat(nm) : pathNames;
+          walkShapes((shapeNode as LottieShapeGroup).it, layerName, nextPath);
+          continue;
+        }
+
+        const segs = pathNames.flatMap(splitSegments);
+        const shapeName = normalize(shapeNode.nm as string | undefined);
+        const pathStr = [...pathNames, shapeNode.nm || '']
+          .filter(Boolean)
+          .join(' / ');
+
+        // 1) Accessories FIRST (so they override hose matches)
+        const matchesSecondary =
+          segs.some(s => anyMatch(s, SECONDARY_RXS)) ||
+          anyMatch(shapeName, SECONDARY_RXS) ||
+          anyMatch(layerName, SECONDARY_RXS) ||
+          anyMatch(pathStr, SECONDARY_RXS);
+
+        const matchesTertiary =
+          segs.some(s => anyMatch(s, TERTIARY_RXS)) ||
+          anyMatch(shapeName, TERTIARY_RXS) ||
+          anyMatch(layerName, TERTIARY_RXS) ||
+          anyMatch(pathStr, TERTIARY_RXS);
+
+        if (matchesSecondary) {
+          paintNode(shapeNode, palette.secondary);
+          continue;
+        }
+        if (matchesTertiary) {
+          paintNode(shapeNode, palette.tertiary);
+          continue;
+        }
+
+        // 2) Rubber-hose lines/areas on arms/legs → PRIMARY (if not accessory)
+        const inArmLegContext =
+          layerSegs.some(isArmLegSeg) || segs.some(isArmLegSeg);
+        const inHoseGroup = segs.some(isHoseySeg);
+        const ty = shapeNode.ty;
+        if ((ty === 'st' || ty === 'fl') && inArmLegContext && inHoseGroup) {
+          paintNode(shapeNode, palette.primary);
+          continue;
+        }
+      }
+    };
+
+    const visitLayer = (layer?: LottieLayer) => {
+      if (!layer) return;
+      if (Array.isArray(layer.shapes)) {
+        walkShapes(layer.shapes, layer.nm || '', []);
+      }
+    };
+
+    animationData.layers?.forEach(visitLayer);
+    animationData.assets?.forEach(a =>
+      (a as LottieAssetPrecomp).layers?.forEach(visitLayer)
+    );
+  }
+
+  /** Loads head.png as a data URL and detects natural size (falls back to 1000×1000). */
```

Comment:
I think it would be much simpler to just hardcode the dimensions of these source assets, rather than waiting on an async image file retrieval just to find out how big it is.